compiler/dyno/lib/resolution/Resolver.cpp

```diff
@@ -2414,6 +2414,91 @@ void Resolver::exit(const IndexableLoop* loop) {
   }
 }
 
+// Returns 'true' if a single Id was scope-resolved, in which case the function
+// will also return via the ID and QualifiedType formals.
+static bool computeTaskIntentInfo(Resolver& resolver, const NamedDecl* intent,
+                                  ID& resolvedId, QualifiedType& type) {
+  auto& scopeStack = resolver.scopeStack;
+
+  // Look at the scope before the loop-statement
+  const Scope* scope = scopeStack[scopeStack.size()-2];
+  LookupConfig config = LOOKUP_DECLS |
+                        LOOKUP_IMPORT_AND_USE |
+                        LOOKUP_PARENTS |
+                        LOOKUP_INNERMOST;
+
+  const Scope* receiverScope = resolver.methodReceiverScope();
+
+  auto vec = lookupNameInScope(resolver.context, scope, receiverScope,
+                               intent->name(), config);
+  if (vec.size() == 1) {
+    resolvedId = vec[0].id(0);
+    if (resolver.scopeResolveOnly == false) {
+      if (resolvedId.isEmpty()) {
+        type = typeForBuiltin(resolver.context, intent->name());
+      } else {
+        type = resolver.typeForId(resolvedId, /*localGenericToUnknown*/ true);
+      }
+    }
+    return true;
+  } else {
+    return false;
+  }
+}
+
+bool Resolver::enter(const ReduceIntent* reduce) {
+
+  ID id;
+  QualifiedType type;
+  ResolvedExpression& result = byPostorder.byAst(reduce);
+
+  if (computeTaskIntentInfo(*this, reduce, id, type)) {
+    result.setToId(id);
+  } else if (!scopeResolveOnly) {
+    context->error(reduce, "Unable to find declaration of \"%s\" for reduction", reduce->name().c_str());
+  }
+
+  // TODO: Resolve reduce-op with shadowed type
+  // E.g. "+ reduce x" --> "SumReduceOp(x.type)"
+  reduce->op()->traverse(*this);
```

Comment:
Do you have ideas about how you are going to do this?

Reply:
If the op-expression is a builtin type, then we'll want to create a suitable type constructor call. Otherwise we should just try to type-resolve the op-expression, I think. Then I believe we try to call the appropriate ``accumulate`` method on the class and use the result as the type of the reduce-intent's shadow variable.